### PR TITLE
#US4R-83: Add verification of measured voltages at the time of setting HV voltage

### PR DESCRIPTION
**Pull request by jrozb91**

### CODE REVIEW

File: arrus/core/devices/us4r/Us4RImpl.cpp [hunk 1]
```diff
@@ -67,6 +67,92 @@ void Us4RImpl::setVoltage(Voltage voltage) {
...
+        measured = getMeasuredPVoltage();
+    }
+    if (abs(measured - static_cast<float>(voltage)) > tolerance) {
+        disableHV();
+        //throw exception
+        throw IllegalStateException(
+            ::arrus::format("Measured HVP voltage on HV module invalid '{}', should be in range: [{}, {}]",
+                measured, (static_cast<float>(voltage) - tolerance), (static_cast<float>(voltage) + tolerance)));
+    }
+    logger->log(LogSeverity::INFO, ::arrus::format("Measured HVP on HV module = {} V", measured));
```
Comment (pjarosik) [inline arrus/core/devices/us4r/Us4RImpl.cpp:98]: Please avoid duplication of code logic here. Note that lines 86-98 could be extracted to a private function e.g.  `checkVoltageValue(float expected, float tolerance, float (Us4RImpl::*getter)(), const std::string& name, uint8_t retries)`. See minimal example [here](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAJz%2BpK4AMngMmAByPgBGmMQgAMzSAA6oCoRODB7evgFBaRmOAmER0SxxCcm2mPbFDEIETMQEOT5%2BgTV1WY3NBKVRsfFJ0gpNLW15nWN9A%2BWVIwCUtqhexMjsHOaJ4cjeWADUJoluTmPEmKzH2CYaAILbu/uYRyfn4cDXtw9mOwx7XkOxzcBAQFyY6C%2B92%2B3z2TAUCgOABFMAA3PAbI4Adis9xSXhitAxIG%2BBzJByotFQTAIB2AmAIADUxF5MHcIItsbi7uTeQcLgQ1gwDqiWZhjhYjvdeSYsUiYdLyZTqbT6UyxRYOVzSXyyQKhSKxUdLAcuAA6DRUCU6smy%2BXQxVk5U0ukM5neTBuLWy7m6vUMg2ij0HAC0B0kFqtiV95LtCp55NRqDw6AOSnVHuoVJdQdZnJ9Nr5AHoi2mGY4jAdQS9c5gzfXC7ySwdIqgAO4HegEMAcRHIBCYZAAawOeCoVYHhuDeD7JAuDilCd1zZiAE8Dv3B0OPgcmBvVil6AdUOP8FQqPFGLSUs1WAz4gozY3yaCZyHrrXXkip6zrY6%2BeEtICsQeCYIixzfgArH%2BS58s6tJEPQxCGJiEEHIkkYwX6G4DsO7qshAtakFWnjxChmDEeYABsKLohsIAgGq%2BFspRZhmJ%2BTDmGYxHAaBCiLFhfqbnhYqEWKxGIWR/wUcaNFohi7CMW6GqsexRoxFxPEMiBYECdGz5ksJQ7MWJHoSaRyHSaxcl0YpTFim4qmfqYbFaQQOn8YJtpyvGKQgaKBDsPGvJJimOFbiZ8EHJgqgpIOgXoMRUWSZZGxJdmtIQLRCkMQAVFQXj/IsHLEc2aSAfEJHIvJmJsKCGAGdhfJoAwYxpgQ6AMe8lbUcwbDEYB/LaXx%2BY4o1UVsPCayYKmaEQK%2BCjvok2D5YVyDFXpMa6m2CB0C8UBMDECgQJNCjTamYYzI4yAAPqiGMwLwdcEAxXFDgzYs%2BbLSRSHkfmZhUdRQ3uXxIYhqNW1NWMnUgAtN2guCMMKPQmApDd/DEBA0MMf2xACKgDEsHQRJKC16DHVwGhU59XlNad51fgc827Yt1yrUVHK0zKPn/ryY5M4dx30xcF3tTSGJ3fCBCPRlz2vfFH1fdgP1SRsEONbyCPtu1MPEIVjhsDd8R45jfUvNYxpmAcDCoLS4RTmFln0lxm2NXGvPktjKCrLSwLAtbd6vG4/tcYzod%2BycBzCzNQf%2B17ri0BKTVNYW7sPP%2BUW1j6GiYT5%2BkOg8efct8g0sEw4TemN/7ZZikL57B6BmumJmQa70Jyhwyy0JwkG8H4HBaKQBMcMHlgWwoqzrObvw8KQBCaJ3yxDiAkhUWagRcJTkhcIkiT%2BFRXBmAAHPonCSH3C9D5wvAKCAGhzwvyxwLASBoCwKR7WQFAQG/H9IYxh9uI0FoIFYgt8IAxEvjEcIzRVycFnm/NgggADyDBaBwIHrwLAZcjDiEwaQfA85HCojApfGKg4vCBUvhVbu%2BCiQxGQsQVcHgsCXxBiweBnc%2BAGGAAoRkoE2zILiv3We/BBAiDEOwKQMhBCKBUOofBuguD6EMMYaw1h9B4COvAZYqAUj1FvhwG%2BtR4pZBcAwdwnh2h6FCOEQYFRhjKMKJkAQkw/BOPSC4hgcwhgJGUXYUxAhegTCsXkfxJiHA9HGP0Ox8xHG2GiW4vQMwWg%2BIcX45YE81j0WUe5TAmweBdx7hffBw9VBHyoiGKikg6TIGQKaMwZorYQFHlYSwxFcCEBIMaRIyiDgeHfp/HpXBFi8Hnpgz6pBl4AyaZIMwkFV5mCxIkKiGgzBcCxFRU%2BHBz6kH7oPYeN874PwmaQZ%2BiAQA%2B3xAQcglBf6f0iHeTg5TKnVNqfUw%2BTTeAzS6SBTqyixHCFEOIaRgK5FqEvko0gbZkIpE4UUjgvc9mX2Hsgyh1zjzjheVUmpwA6kNKaUzAZf9KrbBGWMx%2BS8QAzM2VRZI/hIL%2BDMJIFZkE2XbN2fs3ghzbDHPGVoSZtCzAlIOdfE5ArlgkLAWYyQQA%3D%3D%3D)

**Review by pjarosik — CHANGES_REQUESTED** [review]
Also, to discuss with @mateusz-us4us if user should be able to set the voltage to less than 5V (consider both HV256 and US4RPSC (large platform))

**Comment by jrozb91** [comment]
@pjarosik yes

File: arrus/core/devices/us4r/Us4RImpl.cpp (hunk 1 above)
Comment (jrozb91) [reply arrus/core/devices/us4r/Us4RImpl.cpp]: Duplications removed in later commits

File: arrus/core/devices/us4r/Us4RImpl.cpp [hunk 2]
```diff
@@ -67,6 +67,92 @@ void Us4RImpl::setVoltage(Voltage voltage) {
             ::arrus::format("Unaccepted voltage '{}', should be in range: [{}, {}]", voltage, minVoltage, maxVoltage));
     }
     hv.value()->setVoltage(voltage);
+
+    //Wait to stabilise voltage output
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+    //Verify register
+    if (this->getVoltage() != voltage) {
+        //throw exception
```
[the remark this answers is not in the document]
Comment (jrozb91) [reply arrus/core/devices/us4r/Us4RImpl.cpp]: Yes, I have added exception throw if read register does not match written value

**Comment by jrozb91** [comment]
@pjarosik yes

**Comment by jrozb91** [comment]
@pjarosik yes

**Review by pjarosik — APPROVED** [review]

**Comment by pjarosik** [comment]
@jrozb91 note: the changes have been accepted, please squash and merge changes into v0.8.0-dev if you think the task is complete

**Pull request merged**